### PR TITLE
Update ruby api

### PR DESCRIPTION
**Pull request by ryan-matroid**

### CODE REVIEW

**Comment by ryan-matroid** [comment]
The urls to the full documentation in the README will be updated after merging.

File: lib/matroid.rb [hunk 1]
```diff
@@ -1,43 +1,8 @@
...
-      endpoint = @endpoints[:account_info]
-      headers = { "Authorization" => @authorization_header }
-      opts = {
-        headers: headers
-      }
-      response = @request.send(endpoint[:method], endpoint[:uri], opts)
-      response.parsed_response
-    end
-  end
+  autoload :Detector, 'matroid/detector'
```
Comment (julian-matroid) [inline lib/matroid.rb]: It sounds like autoload will be phased out: http://blade.nagaokaut.ac.jp/cgi-bin/scat.rb/ruby/ruby-core/41149

**Comment by ryan-matroid** [comment]
Also, I am having a hard time using ruby `rest-client` to send multiple files for batch image classification. `req.files` keeps getting sent in with the form
```
{
  file[]: [
    { name: 'image1.jpg', ... },
    { name: 'image2.jpg', ... }
  ]
}
```
unlike with curl where it is `{ file: [...] ...}`.

File: lib/matroid.rb [hunk 2]
```diff
@@ -1,43 +1,75 @@
...
+  class << self
+
+    # Authenticates access for Matroid API
+    # @example
+    #  Matroid.authenticate("<your_client_id>", "<your_client_secret>")
+    # @param client_id [String]
+    # @param client_secret [String]
+    # @return [Boolean] If the the access token is successfully created.
+    def authenticate(client_id = nil, client_secret = nil)
+      return true unless @token.nil? or @token.expired?
```
Comment (julian-matroid) [inline lib/matroid.rb]: `||` is better than `or` in Ruby for boolean logic. `and` and `or` are more for control flow: http://www.virtuouscode.com/2014/08/26/how-to-use-rubys-english-andor-operators-without-going-nuts/

File: lib/matroid/connection.rb [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+BASE_API_URI       = 'https://www.matroid.com/api/0.1'
+DEFAULT_GRANT_TYPE = 'client_credentials'
+TOKEN_RESOURCE     = '/oauth/token'
+VERBS              = %w(get post)
+
+# error messages
+NO_ENV_VARIABLES_ERR     = 'AuthenticationError: "MATROID_CLIENT_ID" and "MATROID_CLIENT_SECRET" not found in environment'
+BAD_ENV_VARIABLES_ERR    = 'AuthenticationError: problem using environment variables "MATROID_CLIENT_ID" and "MATROID_CLIENT_SECRET"'
+BAD_CLIENT_VARIABLES_ERR = 'AuthenticationError: problem using client variables provided'
```
Comment (julian-matroid) [inline lib/matroid/connection.rb]: If you're going to have custom errors, you should subclass them from `StandardError`. The client can react to certain errors differently depending on the type without having to parse the error message.

File: lib/matroid.rb [hunk 4]
```diff
@@ -1,43 +1,75 @@
...
+      if client_id and client_secret
+        raise BAD_CLIENT_VARIABLES_ERR if get_token(client_id, client_secret).nil?
+        @client_id, @client_secret = client_id, client_secret
+      elsif (@client_id.nil? or @client_secret.nil?) and !environment_variables?
+        raise NO_ENV_VARIABLES_ERR
+      else
+        client_id, client_secret = ENV['MATROID_CLIENT_ID'], ENV['MATROID_CLIENT_SECRET']
```
Comment (julian-matroid) [inline lib/matroid.rb]: could be initialized when the module is defined.

File: lib/matroid.rb [hunk 5]
```diff
@@ -1,43 +1,75 @@
...
-      @client_secret = options[:client_secret]
-      @request = HTTParty
+  class << self
+
+    # Authenticates access for Matroid API
+    # @example
+    #  Matroid.authenticate("<your_client_id>", "<your_client_secret>")
+    # @param client_id [String]
+    # @param client_secret [String]
+    # @return [Boolean] If the the access token is successfully created.
```
Comment (julian-matroid) [inline lib/matroid.rb:18]: Comment seems out of date

File: lib/matroid.rb [hunk 6]
```diff
@@ -1,43 +1,75 @@
...
+    #   }
     def account_info
-      endpoint = @endpoints[:account_info]
-      headers = { "Authorization" => @authorization_header }
-      opts = {
-        headers: headers
-      }
-      response = @request.send(endpoint[:method], endpoint[:uri], opts)
-      response.parsed_response
+      get('/account')
```
Comment (julian-matroid) [inline lib/matroid.rb]: nice

File: lib/matroid/connection.rb [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+
+    private
+
+    def get_token(client_id, client_secret)
+      url = BASE_API_URI + TOKEN_RESOURCE
+      params = auth_params(client_id, client_secret)
+      begin
+        response = RestClient.post(url, params)
+      rescue RestClient::ExceptionWithResponse => e
+        puts JSON.pretty_generate(JSON.parse(e.response))
```
Comment (julian-matroid) [inline lib/matroid/connection.rb]: Will this gracefully return an object if the response is a 502 bad gateway with HTML?

File: lib/matroid/connection.rb [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+    end
+
+    def send_request(verb, path, *params)
+      url = path.start_with?('http') ? path : BASE_API_URI + path
+      url, query = *url.split('?')
+      url = URI::encode(url)
+      url << "?#{query}" if query
+
+      # refreshes token with each call
+      authenticate
```
Comment (julian-matroid) [inline lib/matroid/connection.rb]: I think this just sets the `client_id` and `client_secret` as instance variables

File: lib/matroid/connection.rb [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+        @token = nil
+        return false
+      end
+
+      @token = Token.new(JSON.parse(response))
+    end
+
+    def send_request(verb, path, *params)
+      url = path.start_with?('http') ? path : BASE_API_URI + path
+      url, query = *url.split('?')
```
Comment (julian-matroid) [inline lib/matroid/connection.rb]: There could be [more than one question mark](http://stackoverflow.com/questions/2924160/is-it-valid-to-have-more-than-one-question-mark-in-a-url) in the url. [CGI](http://ruby-doc.org/stdlib-2.4.1/libdoc/cgi/rdoc/CGI.html#M000075) might be helpful here.

File: lib/matroid/connection.rb [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+
+
+    # @return [Numeric] Time left before token expires (in seconds).
+    def time_remaining
+      lifetime_in_days = @lifetime / 24.0 / 60 / 60
+      remaining = lifetime_in_days - (DateTime.now - @born)
+      remaining > 0 ? remaining * 24.0 * 60 * 60  : 0
+    end
+
+    def show
```
Comment (julian-matroid) [inline lib/matroid/connection.rb]: if you called this `to_s` and returned a string you'd get a nice pretty print when you call `puts`.

File: lib/matroid/connection.rb [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+
+  end
+
+  # Represents an OAuth access token
+  # @attr [String]   token_type     ex: "Bearer"
+  # @attr [String]   token_str      The actual access token
+  # @attr [DateTime] born           When the token was created
+  # @attr [String]   lifetime       Seconds until token expired
+  class Token
+    attr_reader :born, :lifetime
```
Comment (julian-matroid) [inline lib/matroid/connection.rb]: might be worth having a reader for `:access_token` too

File: lib/matroid/detector.rb [hunk 12]
```diff
@@ -0,0 +1,314 @@
...
+ZIP_FILE_SIZE_LIMIT   = 300 * 1024 * 1024
+module Matroid
+
+
+  # Represents a matroid detector
+  # @attr [String]          id                Detector id
+  # @attr [String]          name              Detector name
+  # @attr [Array<String>]   labels
+  # @attr [String]          permission_level  'private', 'readonly', 'open', 'stock'
+  # @attr [Bool]            owner             is the current authicated user the owner
```
Comment (julian-matroid) [inline lib/matroid/detector.rb]: nice documentation

File: lib/matroid/detector.rb [hunk 13]
```diff
@@ -0,0 +1,314 @@
...
+
+# size limit constants
+IMAGE_FILE_SIZE_LIMIT = 50 * 1024 * 1024
+VIDEO_FILE_SIZE_LIMIT = 300 * 1024 * 1024
+BATCH_FILE_SIZE_LIMIT = 50 * 1024 * 1024
+ZIP_FILE_SIZE_LIMIT   = 300 * 1024 * 1024
+module Matroid
+
+
+  # Represents a matroid detector
```
Comment (julian-matroid) [inline lib/matroid/detector.rb]: capitalize Matroid

File: lib/matroid/connection.rb (hunk 7 above)
Comment (ryan-matroid) [reply lib/matroid/connection.rb]: OOoooooo. Good question.

File: lib/matroid.rb (hunk 5 above)
Comment (ryan-matroid) [reply lib/matroid.rb:18]: What do you mean by that?

Comment (julian-matroid) [reply lib/matroid.rb:18]: The method doesn't create an access token

File: lib/matroid/connection.rb (hunk 8 above)
Comment (ryan-matroid) [reply lib/matroid/connection.rb]: it calls `get_token` if the token is nil or expired, which then sends a request to `/oauth/token`.

File: lib/matroid/connection.rb (hunk 10 above)
Comment (ryan-matroid) [reply lib/matroid/connection.rb]: Ruby power.

File: lib/matroid.rb (hunk 5 above)
Comment (julian-matroid) [reply lib/matroid.rb:18]: Oh, I missed that. 
```
token = get_token(client_id, client_secret)
raise BAD_CLIENT_VARIABLES_ERR if token.nil?
```
is a little clearer.

File: lib/matroid/connection.rb [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+    end
+
+    def authorization_header
+      "#{@token_type} #{@access_token}"
+    end
+
+    # Checks if the current token is expired
+    # @return [Boolean]
+    def expired?
+      lifetime_in_days = @lifetime / 24.0 / 60 / 60
```
Comment (julian-matroid) [inline lib/matroid/connection.rb]: set as an instance variable or calculate this in private method to share more code

File: lib/matroid/detector.rb [hunk 15]
```diff
@@ -0,0 +1,314 @@
...
+  # @attr [String]          id                Detector id
+  # @attr [String]          name              Detector name
+  # @attr [Array<String>]   labels
+  # @attr [String]          permission_level  'private', 'readonly', 'open', 'stock'
+  # @attr [Bool]            owner             is the current authicated user the owner
+  class Detector
+    # HASH { <id> => Detector }
+    @@detectors = {}
```
Comment (julian-matroid) [inline lib/matroid/detector.rb]: It's confusing that a class called `Detector` has a variable inside it named `detectors`. Is this class something more like `DetectorManager`? Haven't read through the rest of the class yet.

File: lib/matroid/connection.rb (hunk 9 above)
Comment (ryan-matroid) [reply lib/matroid/connection.rb]: only the first question mark gets parsed and the following are treated as literal so joining everything as the query string following the first "?" should suffice right?

Comment (julian-matroid) [reply lib/matroid/connection.rb]: For the current use case that works

File: lib/matroid/detector.rb (hunk 15 above)
Comment (ryan-matroid) [reply lib/matroid/detector.rb]: This variable isn't available outside. Think of this class as something like an ORM. The reason for storing the detectors as a hash on the class was to have methods like 
```
Matroid::Detector.find(name: /cat/, state: 'trained')
```
which would search the detectors in storage and return an array of Detector instances.

File: lib/matroid/detector.rb [hunk 16]
```diff
@@ -0,0 +1,360 @@
...
+    #     #         }, {
+    #     #           ...
+    #     #         }
+    #     #       ]
+    #     #     }
+    #     #   ]
+    #     # }
+    def classify_image_url(url)
+      classify('image', url: url)
+    end
```
Comment (julian-matroid) [inline lib/matroid/detector.rb:212]: cool

File: spec/fixtures/Matroid_account_info.yml [hunk 17]
```diff
@@ -0,0 +1,54 @@
+---
```
Comment (julian-matroid) [inline spec/fixtures/Matroid_account_info.yml:1]: use lowercase file name

File: spec/fixtures/Matroid_account_info.yml [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+http_interactions:
+- request:
+    method: get
+    uri: https://ryan.dev.matroid.com/api/0.1/account
```
Comment (julian-matroid) [inline spec/fixtures/Matroid_account_info.yml]: don't want to expose this publicly

File: spec/fixtures/Matroid_account_info.yml [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+      - sameorigin
+      content-security-policy:
+      - script-src 'self' https://cdn.datatables.net https://cdnjs.cloudflare.com
+        http://cdn.segment.com https://www.google-analytics.com http://www.google-analytics.com
+        https://www.youtube.com https://s.ytimg.com https://checkout.stripe.com https://code.ionicframework.com;
+        report-uri /api/report_error
+      etag:
+      - W/"a6-g3D9OWhn58PrnAQybEcGNFLbFx0"
+      set-cookie:
+      - keystone.sid=s%3AA0l4QG_LHt5_7YXSdrkT-xiQwGIxR2fr.w1haZC%2FuhU4hPkYNw1iltJSUpFeExvPVk07xACcUn4s;
```
Comment (julian-matroid) [inline spec/fixtures/Matroid_account_info.yml]: is this a real cookie? let's use fake data for the fixture.

File: spec/fixtures/authenticate_client.yml [hunk 20]
```diff
@@ -0,0 +1,63 @@
+---
+http_interactions:
+- request:
+    method: post
+    uri: https://ryan.dev.matroid.com/api/0.1/oauth/token
+    body:
+      encoding: US-ASCII
+      string: client_id=c7RqWJs9MopTAN0Y&client_secret=55JxPSgh395wRW1IIX1PDLDBeVlUZytI&grant_type=client_credentials
```
Comment (julian-matroid) [inline spec/fixtures/authenticate_client.yml]: security vulnerability if these keys are real

File: spec/lib/matroid/detector_spec.rb [hunk 21]
```diff
@@ -0,0 +1,128 @@
+describe Matroid::Detector do
+  # Keys generated specifically for the tests. Should be removed in the future
```
Comment (julian-matroid) [inline spec/lib/matroid/detector_spec.rb]: 👍 

Use the test endpoints instead

Comment (julian-matroid) [reply spec/lib/matroid/detector_spec.rb]: The person testing this gem should already have client keys and a Matroid account.

File: spec/lib/matroid_spec.rb [hunk 22]
```diff
@@ -0,0 +1,30 @@
...
+    it 'should get correct account info' do
+      credits = account_info['account']['credits']
+      daily = credits['daily']
+      monthly = credits['monthly']
+
+      expect(credits['plan']).to                  eq('premium')
+      expect(credits['concurrentTrainLimit']).to  eq(1)
+    end
+  end
+end
```
Comment (julian-matroid) [inline spec/lib/matroid_spec.rb]: Good job testing

File: matroid.gemspec [hunk 23]
```diff
@@ -30,7 +30,18 @@ Gem::Specification.new do |spec|
...
+
   spec.add_development_dependency "bundler", "~> 1.14"
   spec.add_development_dependency "rake", "~> 10.0"
   spec.add_runtime_dependency 'httparty', "~> 0.14.0"
+  spec.add_development_dependency "minitest"
+  spec.add_development_dependency "vcr"
+  spec.add_development_dependency "fakeweb"
+  spec.add_development_dependency "rspec"
+
+  spec.add_dependency "dotenv"
```
Comment (julian-matroid) [inline matroid.gemspec]: can be removed, `ENV` hash is [built in](https://ruby-doc.org/core-2.2.0/ENV.html)

Comment (ryan-matroid) [reply matroid.gemspec]: "dotenv" loads variables from a `.env` into the environment. So the user can place them in their own `.env` file or outside depending on their needs.

**Comment by ryan-matroid** [comment]
Switched over to using httpclient instead of rest-client because there was a weird thing happening with the request made when classifying a batch.

**Comment by ryan-matroid** [comment]
This won't work until the newer code with the `search_detectors` route is deployed. For testing, you can switch the `base_api_uri` to the dev machine by calling `Matroid.set_base_uri(uri)`.

UPDATE: Works for all of the Detector's classify methods.

File: matroid.gemspec (hunk 23 above)
Comment (julian-matroid) [reply matroid.gemspec]: We should specify in the documentation that we expect them to use `dotenv` then. They may be using `.env` for something else, like Heroku deployment.

Comment (ryan-matroid) [reply matroid.gemspec]: It's in the Readme already :-). Also, this doesn't expect a `.env` file. It just adds the variables from `.env` to `ENV` if they are there. If they are already loaded in the ENV, they will already be in ENV.

Comment (julian-matroid) [reply matroid.gemspec]: Yes, but if they don't know that we will parse their `.env` file into local environment variables, they could run into some subtle bugs. Some people use `.env` for production deployment, but the production variables might not be appropriate in a situation where they're using our API locally. 

It's better not to surprise the users; after we discuss environment variables in `README.md` we should explicitly mention the `dotenv` gem.

Comment (ryan-matroid) [reply matroid.gemspec]: Ah, gotcha. I guess I see your angle.

File: spec/fixtures/Matroid_account_info.yml (hunk 17 above)
Comment (ryan-matroid) [reply spec/fixtures/Matroid_account_info.yml:1]: These are automatically generated after running rspec.

**Comment by ryan-matroid** [comment]
To do: add to rubydoc after merge.

**Comment by julian-matroid** [comment]
Is there a way to get all detectors available for use? 

I got an error trying the `list` method
```
irb(main):009:0> Matroid::Detector.list
NameError: undefined local variable or method `type' for Matroid::Detector:Class
	from /Users/julian/code/api/ruby_matroid/matroid/lib/matroid/detector.rb:77:in `list'
	from (irb):9
	from bin/console:14:in `<main>'
```

**Comment by ryan-matroid** [comment]
@julian-matroid  You can get a list of all user trained detectors by 
```
Matroid::Detector.find(state: 'trained')
```
You can get all the published detectors with 
```
Matroid::Detector.find(published: true)
```

Or published user created detectors: 
```
Matroid::Detector.find(published: true, owner: true)
```
For all the ids you can do something like:
```
Matroid::Detector.find(published: true, owner: true).map{ |detector| detector.id }
```

**Comment by ryan-matroid** [comment]
`list` is only a method to return the cached detectors from previous searches. Maybe it should be changed to "cached".

**Review by julian-matroid — APPROVED** [review]
👍

**Pull request merged**